### ISSUE TITLE
Restricting practitioner access to only DetectedIssue Resource if Patient belongs to certain organization

### ISSUE THREAD

**Issue by nirojshrestha019** [awslabs/fhir-works-on-aws-authz-smart#74]
**What's on your mind?**
A practitioner is given admin access currently but we can remove it by setting `adminAccessTypes = []` for that. Further, I have added a claim in the token as 
`patientOrgsClaim=Organization/<org_id_1>, Organization/<org_id_2>`. 

Now I want to restrict practitioners to have only access to DetectedIssue resource if Patient belongs to certain organization as the following API call:
{{API_URL}}/DetectedIssue?patient:Patient.organization=Organization/<org_id_1>,Organization/<org_id_2>

Any idea if that is possible and what would you suggest looking at to achieve this restriction on the practitioner?

Thanks.

**Versions (please complete the following information):**
 - v3.1.0

**Comment by rsmayda** [comment]
Hmmm, that is an interesting use-case! So out of the box you won't be able to achieve this with this package. This package has good examples of how to achieve this though... 

Lets break this into two parts:
1) Recognizing the new claim
You will need to add this claim to the `userIdentity` object in `verifyAccessToken` good example of this: https://github.com/awslabs/fhir-works-on-aws-authz-smart/blob/mainline/src/smartHandler.ts#L169

2) Creating new authorization scheme:
2a)  Authorizing the requested resource is of `DetectedIssue`: I would add a check to the `verifyAccessToken` method and to `authorizeAndFilterReadResponse` 
2b) Authorizing against the new claim and as an array too: If you look at [`hasAccessToResource` method](https://github.com/awslabs/fhir-works-on-aws-authz-smart/blob/39105fcd7cb94f539d8310e1c4cf3be430c68129/src/smartAuthorizationHelper.ts#L129) you would want to change that to accommodate your new claim and the array.

If you have further questions let me know!

**Comment by nirojshrestha019** [comment]
Hi @rsmayda , thank you for the response and instruction. I am able to add `patientOrgsClaim` in  `userIdentity` object. But I believe accomodating my new claim in `hasAccessToResource` method would not help me filter the DetectedIssue resources I want. I intend to filter DetectedIssue resource using Organization ids in which the patient are referenced. For example here is the API call I want to filter when `patientOrgsClaim` is present: 
` {{API_URL}}/DetectedIssue?patient:Patient.organization=Organization/<org_id_1>,Organization/<org_id_2>`

But the issue here is, DetectedIssue resource doesn't have reference to the organization directly and only has reference to the Patient resource. So I cannot use `hasReferenceToResource` function to filter the DetectedIssue resource with `patientOrgsClaim`.
Please correct me if I am wrong.

Is there any way to intercept the API call to filter the resources returned using the `patientOrgsClaim`? 

One of the alternatives (bad workaround solutions) I could think of is, we can change the API calls in the lambda handler (https://github.com/awslabs/fhir-works-on-aws-deployment/blob/smart-mainline/src/index.ts#L31) entry point event in `fhir-works-on-aws-deployment module`.

Thanks.

**Comment by rsmayda** [comment]
> But the issue here is, DetectedIssue resource doesn't have reference to the organization directly and only has reference to the Patient resource. So I cannot use hasReferenceToResource function to filter the DetectedIssue resource with patientOrgsClaim.
Please correct me if I am wrong.

Ahh okay I missed this piece so you are trying to get many layer deep access control. Currently our reference only uses 1-layer deep. The many layer is difficult since the returned resource does **not** have the direct reference to the claim.

> Is there any way to intercept the API call to filter the resources returned using the patientOrgsClaim?

I would advise forking our smart-authz package and make it fit your needs. One approach is to pass in the Search component into the smart-authz and use the Search to dynamically determine if the requestor has access to the resource

**Comment by nirojshrestha019** [comment]
Hi @rsmayda, Thank you for the reply. But could you please elaborate on which **Search component** are you taking about? I am only aware about [`getSearchFilterBasedOnIdentity`](https://github.com/awslabs/fhir-works-on-aws-authz-smart/blob/mainline/src/smartHandler.ts#L186) function that returns the filter. Thanks.

**Comment by rsmayda** [comment]
> Could you please elaborate on which **Search component** are you taking about?

Sure! What I am thinking is that you can pass in the FHIR Works [search component](https://github.com/awslabs/fhir-works-on-aws-search-es/tree/mainline/src) to the AuthZ component via the [constructor](https://github.com/awslabs/fhir-works-on-aws-authz-smart/blob/mainline/src/smartHandler.ts#L77). Once the search component is in the AuthZ context you could change the methods to use it in the [authorizeAndFilterReadResponse](https://github.com/awslabs/fhir-works-on-aws-authz-smart/blob/mainline/src/smartHandler.ts#L325) or others. Just a thought, you should try to prototype first!

**Comment by nirojshrestha019** [comment]
Thank you so much for your help.

We went with adding an extension in the `DetectedIssue` resource which contains organization references. Then we added,  I am able to add `patientOrgsClaim` in `userIdentity` object. And finally added `"Organization":["extension.valueReference"]` object in  in `"DetectedIssue"` object of [fhirResourceReferencesMatrix.v4.0.1.json](https://github.com/awslabs/fhir-works-on-aws-authz-smart/blob/mainline/src/schema/fhirResourceReferencesMatrix.v4.0.1.json) file. We also removed `"Practitioner"` as [adminAccessTypes](https://github.com/awslabs/fhir-works-on-aws-authz-smart/blob/mainline/src/smartHandler.ts#L81). Thank you again for your help.
